### ISSUE TITLE
suggestion move_fonts_above_scripts ?

### ISSUE THREAD

**Issue by centminmod** [pagespeed/ngx_pagespeed#996]
In terms of visual render time, would it make sense to have a filter for move_fonts_above_scripts similar to move_css_above_scripts ? For locally hosted fonts, it seems the order goes css, js then fonts when I have move_css_above_scripts enabled. 

On webpagetest filmstrip view, you can clearly see speedindex slower as alot of stuff isn't showing on in browser until the locally hosted fonts are processed.

**Comment by jeffkaufman** [comment]
@jmaessen

**Comment by jmaessen** [comment]
I suspect the font order is being determined by the browser – the font specifications end up as additional CSS code.

An interesting caveat here is something like the Google font loader, which is a script that injects CSS onto the page in a browser-dependent way.  For Google fonts in particular there's InlineGoogleFontCss:

https://developers.google.com/speed/pagespeed/module/filter-css-inline-google-fonts

It looks like this is currently running *after* move_css_above_scripts; perhaps simply switching the filter order would help a lot.  I'll try that.

**Comment by centminmod** [comment]
Yeah I have inline_google_font_css enabled but doesn't seem to be working. But I also have locally hosted fonts for other stuff too. The webpagetest filmstrip view looks like this [here](http://www.webpagetest.org/video/compare.php?tests=150730_CC_81fddc7b1ea9be9340536c8de4d2322b-r%3A1-c%3A0&thumbSize=100&ival=100&end=visual)

    Version: 13: on
    
    Filters
    ah      Add Head
    cc      Combine Css
    jc      Combine Javascript
    gp      Convert Gif to Png
    jp      Convert Jpeg to Progressive
    jw      Convert Jpeg To Webp
    mc      Convert Meta Tags
    pj      Convert Png to Jpeg
    ec      Cache Extend Css
    ei      Cache Extend Images
    es      Cache Extend Scripts
    fc      Fallback Rewrite Css
    if      Flatten CSS Imports
    hw      Flushes html
    ci      Inline Css
    ii      Inline Images
    il      Inline @import to Link
    ji      Inline Javascript
    js      Jpeg Subsampling
    rj      Recompress Jpeg
    rp      Recompress Png
    rw      Recompress Webp
    ri      Resize Images
    cf      Rewrite Css
    jm      Rewrite External Javascript
    jj      Rewrite Inline Javascript
    cu      Rewrite Style Attributes With Url
    cp      Strip Image Color Profiles
    md      Strip Image Meta Data
    
    Options
      AdminPath (nap)                              /pagespeed_admin
      AvoidRenamingIntrospectiveJavascript (aris)  True
      BeaconUrl (bu)                               /ngx_pagespeed_beacon
      ConsolePath (ncp)                            /pagespeed_console
      EnableRewriting (e)                          1
      FileCacheInodeLimit (afcl)                   500000
      FileCachePath (afcp)                         /var/ngx_pagespeed_cache
      GlobalAdminPath (ngap)                       /pagespeed_global_admin
      GlobalStatisticsPath (ngsp)                  /ngx_pagespeed_global_statistics
      MessagesPath (nmp)                           /ngx_pagespeed_message
      RewriteLevel (l)                             Core Filters
      StatisticsPath (nsp)                         /ngx_pagespeed_statistics

**Comment by jmaessen** [comment]
Interesting.  I'm seeing the font CSS appear before the DNS prefetch on the
page now – did you change something in your site?  Or is Chrome just
messing with its network prioritization from run to run?

**Comment by centminmod** [comment]
only change is i set

   pagespeed FetchHttps enable;

but i do still see google css requests in webpagetest runs

**Comment by centminmod** [comment]
Looks to be working now for google inline css but the config output still doesn't show google inline css as enabled ?

    Version: 14: on
    
    Filters
    ah  Add Head
    cc  Combine Css
    jc  Combine Javascript
    gp  Convert Gif to Png
    jp  Convert Jpeg to Progressive
    jw  Convert Jpeg To Webp
    mc  Convert Meta Tags
    pj  Convert Png to Jpeg
    ec  Cache Extend Css
    ei  Cache Extend Images
    es  Cache Extend Scripts
    fc  Fallback Rewrite Css 
    if  Flatten CSS Imports
    hw  Flushes html
    ci  Inline Css
    ii  Inline Images
    il  Inline @import to Link
    ji  Inline Javascript
    js  Jpeg Subsampling
    rj  Recompress Jpeg
    rp  Recompress Png
    rw  Recompress Webp
    ri  Resize Images
    cf  Rewrite Css
    jm  Rewrite External Javascript
    jj  Rewrite Inline Javascript
    cu  Rewrite Style Attributes With Url
    cp  Strip Image Color Profiles
    md  Strip Image Meta Data
    
    Options
      AdminPath (nap)                              /pagespeed_admin
      AvoidRenamingIntrospectiveJavascript (aris)  True
      BeaconUrl (bu)                               /ngx_pagespeed_beacon
      ConsolePath (ncp)                            /pagespeed_console
      EnableRewriting (e)                          1
      FetcherTimeOutMs (bfto)                      8000
      FileCacheInodeLimit (afcl)                   500000
      FileCachePath (afcp)                         /var/ngx_pagespeed_cache
      GlobalAdminPath (ngap)                       /pagespeed_global_admin
      GlobalStatisticsPath (ngsp)                  /ngx_pagespeed_global_statistics
      MessagesPath (nmp)                           /ngx_pagespeed_message
      RewriteLevel (l)                             Core Filters
      StatisticsPath (nsp)                         /ngx_pagespeed_statistics
    
    Domain Lawyer
    
    Invalidation Timestamp: (none)

**Comment by centminmod** [comment]
i see what's happening i have pagespeed enabled on a per vhost basis but was looking at main hostname's pagespeed_admin page and not the site's/pagespeed_admin :)

    Version: 14: on
    
    Filters
    ah  Add Head
    cw  Collapse Whitespace
    cc  Combine Css
    jc  Combine Javascript
    gp  Convert Gif to Png
    jp  Convert Jpeg to Progressive
    jw  Convert Jpeg To Webp
    mc  Convert Meta Tags
    pj  Convert Png to Jpeg
    ws  When converting images to WebP, prefer lossless conversions
    di  Delay Images
    ec  Cache Extend Css
    ei  Cache Extend Images
    es  Cache Extend Scripts
    fc  Fallback Rewrite Css 
    if  Flatten CSS Imports
    hw  Flushes html
    ci  Inline Css
    gf  Inline Google Font CSS
    ii  Inline Images
    il  Inline @import to Link
    ji  Inline Javascript
    js  Jpeg Subsampling
    ll  Lazyload Images
    cj  Move Css Above Scripts
    cm  Move Css To Head
    pr  Prioritize Critical Css
    rj  Recompress Jpeg
    rp  Recompress Png
    rw  Recompress Webp
    ri  Resize Images
    cf  Rewrite Css
    jm  Rewrite External Javascript
    jj  Rewrite Inline Javascript
    cu  Rewrite Style Attributes With Url
    is  Sprite Images
    cp  Strip Image Color Profiles
    md  Strip Image Meta Data
    
    Options
      AdminPath (nap)                                 /pagespeed_admin
      AvoidRenamingIntrospectiveJavascript (aris)     True
      BeaconUrl (bu)                                  /ngx_pagespeed_beacon
      ConsolePath (ncp)                               /pagespeed_console
      CssImageInlineMaxBytes (cii)                    0
      CssInlineMaxBytes (ci)                          8192
      EnableRewriting (e)                             1
      FetcherTimeOutMs (bfto)                         8000
      FetchHttps (fhs)                                enable
      FileCacheInodeLimit (afcl)                      500000
      FileCachePath (afcp)                            /var/ngx_pagespeed_cache
      GlobalAdminPath (ngap)                          /pagespeed_global_admin
      GlobalStatisticsPath (ngsp)                     /ngx_pagespeed_global_statistics
      ImageInlineMaxBytes (ii)                        4096
      ImageResolutionLimitBytes (irlb)                16777216
      InPlaceResourceOptimization (ipro)              False
      JpegRecompressionQualityForSmallScreens (iqss)  60
      LogDir (ald)                                    /var/log/pagespeed
      MaxCombinedCssBytes (xcc)                       -1
      MaxCombinedJsBytes (xcj)                        135000
      MaxInlinedPreviewImagesIndex (mdii)             -1
      MemcachedServers (ams)                          localhost:11211
      MemcachedThreads (amt)                          1
      MemcachedTimeoutUs (amo)                        100000
      MessagesPath (nmp)                              /ngx_pagespeed_message
      MinImageSizeLowResolutionBytes (nislr)          3072
      PreserveUrlRelativity (pur)                     True
      RewriteDeadlinePerFlushMs (rdm)                 50
      RewriteLevel (l)                                Core Filters
      Statistics (ase)                                True
      StatisticsLogging (asle)                        True
      StatisticsPath (nsp)                            /ngx_pagespeed_statistics
      WebpRecompressionQualityForSmallScreens (iwss)  60
      XHeaderValue (xhv)                              centminmod.com PageSpeed

**Comment by centminmod** [comment]
FYI, i saved the google fonts locally to be served and seems the fonts are pushed up higher in the order and speeds up [speedindex by 0.2-0.3s](http://www.webpagetest.org/video/compare.php?tests=150731_MQ_1a3e8d0f6bb7d8e9d773d624dd3d7275-r%3A1-c%3A0&thumbSize=100&ival=100&end=visual)

**Comment by jmaessen** [comment]
On Thu, Jul 30, 2015 at 11:19 PM, George Liu (eva2000) <
notifications@github.com> wrote:

> FYI, i saved the google fonts locally to be served and seems the fonts are
> pushed up higher in the order and speeds up speedindex by 0.2-0.3s
> <http://www.webpagetest.org/video/compare.php?tests=150731_MQ_1a3e8d0f6bb7d8e9d773d624dd3d7275-r%3A1-c%3A0&thumbSize=100&ival=100&end=visual>
>

Careful, though – part of the point of Google Fonts (if you stick with the
popular ones) is to have a repeat-view experience for a first-time visitor
to your site.

-Jan

**Comment by centminmod** [comment]
yeah @jmaessen will have to keep an eye of my google analytics page speed stats for first time and repeat visitors :)
